### PR TITLE
Feature: Instance name func to doc file

### PR DESCRIPTION
**Pull request by Ferrany1**
**Describe the PR**
Feature: Instance name func to doc file

**Relation issue**
https://github.com/swaggo/swag/issues/1113

### CODE REVIEW

File: gen/gen.go [hunk 1]
```diff
@@ -414,31 +414,37 @@ import (
...
+	instanceName: {{ printf "%q" .InstanceName }},
 }
 
-type s struct{}
+type s_{{ .InstanceName }} struct{}
 
-func (s *s) ReadDoc() string {
-	sInfo := SwaggerInfo
+func (s *s_{{ .InstanceName }}) ReadDoc() string {
+	sInfo := SwaggerInfo_{{ .InstanceName }}
```
Comment (ubogdan) [inline gen/gen.go:447]: Why do we need an additional s_InstanceName struct when we can define the ReadDoc() for swaggerInfo_InstanceName?
```
func (i *swaggerInfo_{{ .InstanceName }}) ReadDoc() string { 
```

**Review by ubogdan — CHANGES_REQUESTED** [review]

File: gen/gen.go (hunk 1 above)
Comment (Ferrany1) [reply gen/gen.go:447]: How are looking on defining swaggerInfo struct inside swag as exported struct and generating only var SwaggerInfo_InstanceName with all methods?

Comment (ubogdan) [reply gen/gen.go:447]: replace
```
func (s *s_{{ .InstanceName }}) ReadDoc() string {
```

with 
```
func (s *swaggerInfo_{{ .InstanceName }}) ReadDoc() string { 
```

remove
```
type s_{{ .InstanceName }} struct{}
sInfo := SwaggerInfo_{{ .InstanceName }}
```

update 
```
	if err := t.Execute(&tpl, sInfo); err != nil {
```
to 
```
	if err := t.Execute(&tpl, s); err != nil {
```

Comment (ubogdan) [reply gen/gen.go:447]: Does that make sense now ?

Comment (Ferrany1) [reply gen/gen.go]: How about doing it this way, it wont break any current cases

Comment (ubogdan) [reply gen/gen.go]: Removing ReadDoc() function will introduce breaking changes. 

See yourself by regenerating the documentation for example/basic and building the project with the proposed way.

Comment (Ferrany1) [reply gen/gen.go]: I've tried to, with importing local Swagger info struct, worked fine

Comment (Ferrany1) [reply gen/gen.go]: I'm referring to swag.SwaggerInfo, which matches Swagger interface

Comment (Ferrany1) [reply gen/gen.go]: @ubogdan what do you think about changes?

Comment (ubogdan) [reply gen/gen.go]: @Ferrany1 congratulations! moving the code from string constant to swag enable us to properly write unit tests against it. 
I'm quite caught with different things, but I will try to do a CR soon.

File: swagger_info.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	"strings"
+	"text/template"
+)
+
+// DocumentationReader is an interface to inject Swagger info reader.
+type DocumentationReader interface {
+	Swagger
+	InstanceName() string
+}
+
```
Comment (ubogdan) [inline swagger_info.go:15]: A good practice is to define the interface where you use it not where you implement it. Please remove this definition ...

File: gen/gen.go [hunk 3]
```diff
@@ -405,68 +405,23 @@ var packageTemplate = `// Package {{.PackageName}} GENERATED BY THE COMMAND ABOV
...
-	Schemes     []string
-	Title       string
-	Description string
-}
-
-// SwaggerInfo holds exported Swagger Info so clients can modify it
-var SwaggerInfo = swaggerInfo{
+// SwaggerInfo_{{ .InstanceName }} holds exported Swagger Info so clients can modify it
+var SwaggerInfo_{{ .InstanceName }} = &swag.SwaggerInfo{
 	Version:     {{ printf "%q" .Version}},
```
Comment (ubogdan) [inline gen/gen.go:414]: Please rename the struct to Doc so the definition will not contain the swag string to times.
```
var SwaggerInfo_{{ .InstanceName }} = &swag.Spec {
```

File: swagger_info.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	Version          string
+	Host             string
+	BasePath         string
+	Schemes          []string
+	Title            string
+	Description      string
+	InfoInstanceName string
+	SwaggerTemplate  string
+}
+
```
Comment (ubogdan) [inline swagger_info.go:27]: Please rename the struct to Spec and also the filename spec.go

**Review by ubogdan — CHANGES_REQUESTED** [review]
Good job till now. Would you mind making some other minor adjustments according to the comments?

File: swagger_info.go (hunk 4 above)
Comment (Ferrany1) [reply swagger_info.go]: done

File: gen/gen.go (hunk 3 above)
Comment (Ferrany1) [reply gen/gen.go:414]: Done

File: swagger_info.go (hunk 2 above)
Comment (Ferrany1) [reply swagger_info.go]: done

**Review by ubogdan — APPROVED** [review]
LGTM

**Review by ubogdan — APPROVED** [review]
LGTM

**Pull request merged**